오픈마켓 [STEP 2] 알라딘, 조이

na-young-kwon opened the pull request:
안녕하세요 그린! **[@GREENOVER](https://github.com/GREENOVER)🫒**

3조 알라딘과 조이입니다

이번에 `View` 만들면서 이 방법 저 방법 고민하면서 해봤는데

수정이 필요한 부분들 한번씩 짚어주시면 감사하겠습니다 🙂

이번 리뷰도 잘 부탁드립니다~~ 🙏

# 고민했던 점

## View를 구성하는 방법

지금 되어있는 방법으로 구현하기 전에 다른 방법으로 뷰를 만들려고 시도해봤습니다.

하나의 ViewController (`MarketViewController`) 에서 `SegmentedControl` 을 toggle하면 `ListCell` , `GridCell` 만 교체하는 방법으로 시도해봤습니다.

그런데 여기서 발생했던 문제는, `SegmentedControl` 사이를 왔다갔다 할때마다 `ListCell` 이랑 `GridCell` 이 섞여서 표시된다는 것이었습니다. 

<img src="https://user-images.githubusercontent.com/74536728/149463289-f705456e-54c7-4344-b14e-995989c8a268.gif" width="50%" height="50%">

`reloadView()` 를 적절하게 이용해서 최대한 해결해보려고 했는데, 매끄럽게 전환이 안됐습니다. 

이 문제가 `reuseQueue` 에 `ListCell` 그리고 `GridCell` 이 섞여 들어가면서 발생한거 같은데 결국 해결하지 못했습니다.

그래서 저희가 선택한 방법은 `MarketViewController` 가 `ListViewController` 와 `GridViewController` 를 왔다 갔다 할 수 있도록 구현한 것이었습니다. 

그리고 왔다 갔다 할때마다 뷰컨을 생성하는건 비효율적인 방법 같아서 다음과 같이 지연 속성으로 구현해봤습니다.

```swift
private lazy var listViewController: ListViewController = {
...
}()

private lazy var gridViewController: GridViewController = {
...
}()
```

# 신경쓴 점

1. Loading Indicator의 위치

스토리보드로 구현 시 로딩 인디케이터의 위치가 중앙보다 살짝 아래에 있어서,  중앙정렬 할 수 있도록 코드로 구현했어요

```swift
private lazy var loadingIndicator: UIActivityIndicatorView = {
    let loadingIndicator = UIActivityIndicatorView(style: .large)
    loadingIndicator.isHidden = false
    
    self.view.addSubview(loadingIndicator)
    loadingIndicator.translatesAutoresizingMaskIntoConstraints = false
    loadingIndicator.centerXAnchor.constraint(equalTo: self.view.centerXAnchor).isActive = true
    loadingIndicator.centerYAnchor.constraint(equalTo: self.view.centerYAnchor).isActive = true
    
    return loadingIndicator
}()
```

이렇게 구현해놓고 적절한 곳에서 `.startAnimating()` 과 `.stopAnimating()` 으로 사용했습니다.

### 문제상황들

문제 1. 이름이 길어졌을 때 잔여수량 레이블과 겹쳐지는 문제

해결 방법: 두 스택뷰 사이에 anchor를 걸어서 해결해주었습니다


<img src="https://i.imgur.com/r4pvyeh.png" width="50%" height="50%">


→

`문제 1` 을 해결했더니 다른 문제가 생겼어요

문제상황 2. 이름 레이블이 (...)으로 표시되는 문제

해결 방법: 두 스택뷰 간의 anchor를 걸어서 10정도 간격을 유지하도록 했어요
<img src="https://i.imgur.com/lsKyPPo.png" width="50%" height="50%">




`문제 2`를 해결했더니

문제 3. 가격 레이블이 (...) 으로 표시되는 문제
<img src="https://i.imgur.com/SLy7qiy.png" width="50%" height="50%">


해결 방법:  nameLabel, stockLabel, disclosureIndicator를 하나의 스택으로 만들고 `compression resistence priority` 를 조정해주었어요.  각각 751, 750, 750으로 설정하여 글자가 길어졌을 때 제일 중요하다고 생각되는 `제품이름` 이 자신의 크기를 유지하도록 설정했습니다

<img src="https://i.imgur.com/329mw9h.jpg" width="50%" height="50%">


### 최종
<img src="https://i.imgur.com/i1gX2xk.png" width="50%" height="50%">



가격 레이블이 재고 레이블 만큼 길어져도 (...)으로 표시되지 않습니다

### 1. 로딩 인디케이터의 애니메이션을 끝내는 시점

로딩 인디케이터의 애니메이션을 언제 끝내야 할 지 고민되었습니다. 아래 코드에서 1,2,3번 중에 저희는 1번을 선택했어요. 이유는 인디케이터를 멈춘 후 성공/실패 여부에 따라 적절한 행동을 해야한다고 생각했기 때문입니다.

2번에 있으면 통신에 실패했을 때 로딩인디케이터의 애니메이션을 끝낼 수 없기 때문에 부적절 하다고 생각했습니다.

3번이 부적절하다고 생각한 이유는 만약 통신에 실패한다면 인디케이터의 애니메이션을 끝내는게 아니라 alert을 보여줘야 한다고 생각했어요.

저희가 옳게 선택했는지 궁금합니다.

```swift
private func fetchPage() {
        startLoadingIndicator()
        apiService.fetchPage(pageNumber: 1, itemsPerPage: 20) { [weak self] result in
            self?.stopLoadingIndicator() // 1번
            switch result {
            case .success(let data):
            self?.stopLoadingIndicator() // 2번
                self?.products = data.products
                self?.showListViewController()
            case .failure(let error):
                print(error)
            }
            self?.stopLoadingIndicator() // 3번
        }
    }
```

### 2. Cell에 데이터를 전달하는 방법

- List / Grid cell을 configure할 때 데이터를 어떻게 전달할 지 고민했어요
- 저희는 메서드의 파라미터로 받아오는 방법을 선택했는데, 이유는 cell의 프로퍼티에 직접 접근하는 것 보다 메서드로 (인스턴스에게 일을 시켜서) 전달하는것이 더 객체지향적이라고 생각했기 때문이에요.

전) cell의 product 프로퍼티에 직접 product를 넣어주는 방법

```swift
func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
    guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CellIdentifier.list, for: indexPath) as? ListCell else {
        return UICollectionViewCell()
    }
    cell.products = products[indexPath.row]
    
    return cell
}
```
후) 메서드의 파라미터로 전달하는 방법 

```swift
func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
    guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CellIdentifier.list, for: indexPath) as? ListCell else {
        return UICollectionViewCell()
    }
    cell.configure(with: products[indexPath.row])
    
    return cell
}
```

# 조언을 얻고싶은 부분

### DeploymentTarget 설정 관련

- 컬렉션뷰의 cell을 list 형식으로 보여주기 위해 `UICollectionLayoutListConfiguration`를 사용했습니다. iOS 14.0이상 버전에서부터 사용할 수 있기 때문에 앱의 deploymentTarget을 14.1로 지정해주었어요. 만약 14이하의 버전에서 앱을 사용해야 한다면 어떻게 대응을 해야할지 궁금해졌어요.

## String extension 활용

`“ ”` 를 사용해야되는 경우에 어떤 방법으로 문자열이랑 붙여줄지에 대한 고민을 해봤습니다.

결국 `Namespace` 를 만들어서 실제로 구현된 방법은 조금 다르지만, 아래와 같은 코드를 놓고봤을때는

어떤 방법이 더 가독성이 좋은지 고민되었습니다. 그린의 의견도 궁금합니다~

(`UILabel` 에 `KRW 10000` 같은 문자열을 넣기 위한 코드입니다)

```swift
priceLabel.text = priceString.prefix(with: currency, separatedBy: " ")
```

```swift
priceLabel.text = currency + " " + priceString
```

## Namespace vs String Extension

`Namespace` 문자열을 만들때 다음과 같이 `“잔여수량: “` 에서 `잔여수량` , `:` , `" "` 를 분리해야되는지에 대한 고민을 했습니다.

```swift
enum CellString {
	static let space = " "
	static let remainingStock = "잔여수량: "
}
```

띄어쓰기는 다음과 같이 `extension` 으로 처리하는게 좋을지도 고민 해봤습니다.

```swift
extension String {	
	var withSpace: String {
		return self + " "
	}
} 
```

결국 첫 번째 방법처럼 문자열이 쓰이는 그대로 `namespace` 를 만드는게 더 직관적일것 같아서 그렇게 했는데

`Green` 의 의견도 궁금합니다!

## MarketAPIService 의존성 주입

`MarketViewController` 가 `MarketAPIService` 에 대한 의존성을 갖고있는 상태인데 의존성 주입을 위해서 `MarketViewController` 에 생성자를 만들어줘야되는지에 대한 고민을 해봤습니다.

그렇게되면 `MarketAPIService` 의 주입을 `SceneDelegate` 에서 해주거나, `default init` 으로 넣어줘야될것 같은데 이에 대한 그린의 생각이 궁금합니다!

또한 현업에서는 어떻게 해결하는지도 여쭤보고싶습니다 :)

junbangg committed:
refactor: MarketAPIService에 하드코딩된 statusCode 수정 #1

na-young-kwon committed:
feat: 상품목록화면 ListCell 구현 #3

- model, view, controller로 폴더링
- ListCell xib 구현
- collectionView 추가
- ViewController -> MarketViewController로 네이밍변경

junbangg committed:
feat: MarketViewController에 CollectionView 구현 #3 #4

- ListCell 오토레이아웃 수정
- ListCell에 configure 메서드 추가
- collectionView의 datasource 구현
- 통신한 데이터 CollectionView list에 반영

na-young-kwon committed:
feat: GridCell 및 segmentedControl 구현 #3

- GridCell xib 구현
- GridCell 커스텀 클래스 추가
- segmentedControl 액션 메서드 구현

junbangg committed:
feat: Grid 형태의 CollectionView 구현 #3 #4

- UICollectionViewDelegateFlowLayout 프로토콜 메서드 구현
- PresentationMode 열거형으로 grid, list 형태 구분
- cellForItemAt 메서드 grid, list 에 맞게 구현
- segmentedControl이 눌리면 버그 발생

na-young-kwon committed:
feat: GridCell 관련 설정 추가 #5

- GridCell의 border 추가
- 레이블 가운데 정렬
- 양 옆 간격을 주기 위해 inset 추가

junbangg committed:
feat: List와 Grid CollectionView를 두개의 ViewController로 분리 #5

na-young-kwon committed:
feat: GridViewController 구현 #5

- 스토리보드에 GridView 추가
- GridViewController 구현
- products를 주입받기 위해 List/Grid 뷰컨트롤러에 이니셜라이저 추가
- 비슷한 성격의 메서드 extension으로 구분

junbangg committed:
feat: SegmentedControl 색 설정 #5

na-young-kwon committed:
feat: LoadingIndicator 추가 #5

junbangg committed:
feat: ListCell 세부 로직 구현 #5

- ActivityIndicator 화면 중앙 정렬

na-young-kwon committed:
refactor: GridCell 세부 로직 구현 #5

- ListCell 및 GridCell 오토레이아웃 수정

junbangg committed:
refactor: 이름, 재고 label이 겹치지않도록 오토레이아웃 수정 #5

na-young-kwon committed:
refactor: 전체적인 리팩토링 #5

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -9,13 +9,20 @@
 		C70FB0FB25BEF61C00C9924E /* AppDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FA25BEF61C00C9924E /* AppDelegate.swift */; };
 		C70FB0FD25BEF61C00C9924E /* SceneDelegate.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FC25BEF61C00C9924E /* SceneDelegate.swift */; };
-		C70FB0FF25BEF61C00C9924E /* ViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FE25BEF61C00C9924E /* ViewController.swift */; };
+		C70FB0FF25BEF61C00C9924E /* MarketViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70FB0FE25BEF61C00C9924E /* MarketViewController.swift */; };
```
명확한 뷰컨 네이밍 너무 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("class does not support nscoder")
```
fatalError로 앱 크래쉬를 내는거보다 더 나은 방법은 혹시 없을까요?🤔

na-young-kwon replied:
저희가 생각해본 방법은

1. 빈배열로 products 초기화 하기
```swift
required init?(coder: NSCoder) {
        self.products = []
        super.init(coder: coder)
    }
```

2. assertionFailure + nil 리턴
```swift
required init?(coder: NSCoder) {
        assertionFailure("init(coder:) has not been implemented")
        return nil
    }
```
입니다.

찾아보니 2번 방법은 앱 크래쉬를 내지 않아서 좀 더 안전한 방법이라고 하는데
어떤 방법이 더 적절할까요?

GREENOVER replied:
네! 2번 방향이 현재보다 훨씬 더 안전하고 적절할것 같아요ㅎㅎ

junbangg replied:
혹시 그린도 2번 방법으로 사용하시나요???

GREENOVER replied:
엄밀히보면 2번 방향에 가깝다고 볼 수 있겠어요!
단지 데이터가 없다해서 화면을 그려주지 못하고 끝나는게 아닌 후속 처리들을 해주고
리포팅까지 해주는 그런것들을 심어 구현하긴 합니다.

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+import UIKit
+
+final class GridViewController: UIViewController {
+    //MARK: - IBOutlets
```
주석들을 달아주셔서 가독과 탐색이 너무 편하네요!👍

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+
+extension GridViewController {
+    private func configureCollectionViewGrid() {
+        let layout = UICollectionViewFlowLayout()
```
`UICollectionViewCompositionalLayout`과 `UICollectionViewFlowLayout`은 어떤차이가 있어서
리스트와 그리드의 레이아웃을 다른 타입으로 가져갔을까요?🙋🏻

junbangg replied:
저희는 원래 하나의 컬렉션 뷰에서 `cell` 만 갈아끼워주는 방법으로 시도해봤습니다.
그 방법이 실패하고나서 두개의 `controller` 를 보여주는 방법으로 바꾼건데, 바꾸는 과정에서 `UICollectionViewFlowLayout` 으로도 `list` 를 만들 수 있는줄을 몰랐습니다.
그래서 어제 한 번 해봤는데, `cell` 간의 `border`? 이 없더라고요???🤔
그래서 `tableView` 처럼 만드는 방법에 대해서도 고민하다가, 이럴거면 그냥 `tableView` 사용하는게 낫지않나?라는 생각을 했습니다.
혹시 그린의 생각은 어떠신가요???

GREENOVER replied:
아하 셀 간 separator line같은것들은 다 커스텀하게 만들어줘야할거에요.
사실 테이블뷰에서 제공하는 기본적인것들이 다 커스텀하게 구현해주긴 해야되서 손이 많이 가긴합니다ㅠ
그럼 테이블뷰가 더 낫지 않나..? 라는 생각이 들텐데,
정리해보면 테이블뷰로 할 수 있는 모든건 컬렉션뷰로 다 구현이 가능합니다.
위에서 말한 손이 많이 가긴하지만, 대신 테이블뷰에서는 제약사항이 많기에 일부 회사들에서는 아예 컬렉션뷰로 다 대체하는곳도 많더라구요.
이에 리스트 화면을 둘중 어느것으로 구현할지는 정말 스타일과 선택의 차이입니다.
다만 그 선택하는 기준이 어렵고 안되서가 아닌 명확한 기준이 있다면 테이블뷰로의 구현도 찬성입니다👍

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        setupCollectionViewCells()
+        configureCollectionViewGrid()
```
set과 configure의 네이밍을 붙일때 어떤 기준으로 선택해주시는지 궁금합니다🙋🏻

na-young-kwon replied:
네이밍의 기준은 setup은 초기세팅과 관련된 작업일 때, configure - UI속성들을 설정할 때 입니다

---
저희가 생각해도 헷갈리는 네이밍인 것 같아서 
```swift
setupCollectionViewCells()
setupGridLayout()
```
으로 변경했습니다 :)

GREENOVER replied:
좋습니다! 잘 차이를 보신것 같아요ㅎㅎ
setXXX는 받아온 값으로 아예 초기세팅 즉 기존 값은 제거하고 전달한 값으로 새롭게 쓰는 동작을 의미할때 저는 사용합니다.
이 맥락에서도 이상없다고 보입니다👍

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CellIdentifier.grid, for: indexPath) as? GridCell else {
+            return UICollectionViewCell()
+        }
```
이렇게 파라미터가 길어지고 붙을 경우가 많기에 저는 아래와 같은 코드 컨벤션을 가져가는걸 추구합니다🙋🏻
사실 많은 다른 코드들도 이런 코드 컨벤션을 가져가는것으로 알고 있구요..!
이런것도 있다~ 라고 봐주시면 될것 같습니다ㅎㅎ
```swift
guard let cell = collectionView.dequeueReusableCell(
    withReuseIdentifier: CellIdentifier.grid, 
    for: indexPath
) as? GridCell else {
    return UICollectionViewCell()
}
```

junbangg replied:
저희도 그런 방법으로 다시 바꿔봤습니다!

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        let width = collectionView.frame.width / 2 - 15
+        let height = UIScreen.main.bounds.height / 3
```
이 수치는 어떤 기준으로 정해주셨는지 궁금합니다~🙋🏻

na-young-kwon replied:
1. 너비는 collectionView 기준 / 높이는 UIScreen 기준인 이유
처음에 `let height = collectionView.frame.height / 3` 이렇게 컬렉션뷰의 사이즈를 이용해서 높이를 설정했었는데 아이폰8같은 홈버튼이 있는 기기에서 글자들이 겹치는 현상이 있었어요.
그래서 `UIScreen.main.bounds.height / 3` 처럼 스크린 사이즈로 계산을 해주었던 것이었습니다. 

디버깅을 하다가 나누는 값(3)이 너무 커서 이런 현상이 일어난다는것을 알았고, 
```swift
let width = collectionView.frame.width / 2 - 15
let height = collectionView.frame.height / 2.5
```
둘 다 컬렉션뷰를 기준으로 수치를 정하는 대신 높이를 나누어주는 값을 3 -> 2.5로 변경했습니다.

2. 수치를 정한 기준
빌드해보면서 명세에 나와있는 사진과 가장 비슷해보였기 때문이에요~!
셀의 너비는 컬렉션뷰의 반보다 15정도 작게, 높이는 화면에 셀이 2.5~3개 정도 들어간다고 생각하고 정했습니다.

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -0,0 +1,95 @@
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
+        return UIEdgeInsets(top: 0, left: 10, bottom: 0, right: 10)
```
leading과 left, trailing과 right 다 똑같은것 같은데 어떤 차이가 있을까요?🙋🏻
아마 아실수도 있지만 이 부분은 아주 잘 애플의 철학이 녹아져 있습니다.👍

junbangg replied:
`leading`, `trailing` 은 글이 시작되는 방향이고
`left`, `right` 는 그냥 화면 기준에서의 방향인걸로 알고있습니다.

애플에서 이런 차이를 둔것은 아랍어처럼 글자가 오른쪽-> 왼쪽으로 읽히는 언어들이 있기 때문인걸로 알고있습니다.

GREENOVER replied:
💯💯💯 퍼펙트합니다!

GREENOVER commented on OpenMarket/OpenMarket/ListViewController.swift:
```diff
@@ -0,0 +1,71 @@
+        
+        setupCollectionViewCells()
+        configureCollectionViewList()
+        collectionView.dataSource = self
```
리스트뷰에서는 delegate를 채택하지 않아도 되는 이유가 어떤걸까요?🙋🏻

na-young-kwon replied:
리스트 뷰 에서는 grid뷰와 달리 `UICollectionViewDelegateFlowLayout`을 사용하지 않기 때문이라고 생각합니다~!

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        static let mainStoryBoard = "Main"
+        static let list = "ListViewController"
+        static let grid = "GridViewController"
+    }
```
스토리보드와 뷰컨트롤러가 같은 타입으로 묶이는게 적절할까요?🤔

junbangg replied:
저희가 스토리보드는 빼는게 좋을것 같다고 생각했습니다.
저희가 생각하는 "namespace 를 만드는 기준" 은 크게

1. 한곳에서 관리할 수 있게
2. 자주 재사용되는 문자열들을 매번 생성되지 않게 하기위해

그런데 위 기준들을 생각해보면 `"Main"` 같은 경우에는 따로 타입을 만들 필요가 있나?라는 생각이 들었습니다.
따로 타입을 만들기에는 연관된 다른 값들이 없을것 같고, `ViewController` 가 만들어지면서 한번씩만 생성되기 때문에 그냥 다음과 같이 "Main" 을 `private` 상수로 만들어서 쓰는게 어떨까?라는 생각을 해봤습니다.


```swift
private let mainStoryboardName = "Main"
...
private lazy var listViewController: ListViewController = {
    let storyboard = UIStoryboard(name: mainStoryboardName, bundle: Bundle.main)
    let viewController = storyboard.instantiateViewController(identifier: ListViewController.identifier) { coder in
        ListViewController(products: self.products, coder: coder)
    }
    return viewController
}()
```

혹시 그린이 생각하시는 "namespace" 를 사용하는 기준들이 추가적으로 더 있는지, 그리고 "Main" 문자열을 관리하기 위한 더 좋은 방법이 더 있을지 궁금합니다 :)

GREENOVER replied:
저는 다른 맥락으로 질문을 드렸던 부분인것 같아요..!
저는 namespace를 같은 범주에 있는 애들끼리 묶고 있어요.
예를들어 위에서는 컨트롤러들의 네임 즉 list,grid는 같은 범주라고 봤는데 스토리보드는 성질이 조금 다르지 않나 싶었어요.
그래서 Identifier로 접근한다면 이 안에 컨트롤러들만 들어있어야할것 같은데 스토리보드도 있네?라고 생각이 들었던것 같습니다🙋🏻
스토리보드가 추후 여러개가 될 수 있다면 스토리보드끼리 namespace 타입을 만들어 관리하는것도 방법일것 같습니다~!
코멘트로 수정해주신 방향도 우선 컨트롤러들이 담긴 namespace에서 빠져 좋다고 봅니다!
결과적으로 추후 만약 이 스토리보드도 namespace로 관리해주고 싶다면 또 다른 하나의 namespace를 만드는게 어떨까해요.

junbangg replied:
아! storyboard를 여러개 사용하는것에 대해서는 생각을 못해봤네요!
말씀하신대로 storyboard namespace 하나 만들어볼게요!!

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        return viewController
+    }()
+    
+    private lazy var loadingIndicator: UIActivityIndicatorView = {
+        let loadingIndicator = UIActivityIndicatorView(style: .large)
+        loadingIndicator.isHidden = false
+        
+        self.view.addSubview(loadingIndicator)
+        loadingIndicator.translatesAutoresizingMaskIntoConstraints = false
+        loadingIndicator.centerXAnchor.constraint(equalTo: self.view.centerXAnchor).isActive = true
+        loadingIndicator.centerYAnchor.constraint(equalTo: self.view.centerYAnchor).isActive = true
+        
+        return loadingIndicator
+    }()
```
로딩 인디케이터 구현 너무 좋네요!👍
사용자를 위한 세심한 배려에 감동받고 갑니다👍

na-young-kwon replied:
구우웃👍🏻

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        super.viewDidLoad()
+        
+        setupSegmentedControl()
+        fetchPage(pageNumber: 1, itemsPerPage: 20)
```
아마 요구사항에는 없지만 추후 pagination의 구현이 필요한 경우가 올것 같습니다~!
지금 서버에 무수히 많은 데이터들이 있는데 20개밖에 못받아와지는 한계가 있기때문이죠🥲

na-young-kwon replied:
저희도 몇개의 아이템을 가져와야 하는지 고민하다가 현재는 20개를 가져오고 있는데
추후에 pagination을 구현하도록 리팩토링이 필요할 것 같네요~!

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -0,0 +1,136 @@
+extension MarketViewController {
+    @IBAction func segmentedControlTapped(_ sender: UISegmentedControl) {
+        if sender.selectedSegmentIndex == 0 {
+            remove(asChildViewController: gridViewController)
```
숨길수도 있고 선택되었을때 해당뷰를 최상단으로 올리는것으로도 가능할텐데
매번 지웠다 생성했다 해주는것으로 채택하신 이유가 있을까요?🙋🏻
(잘못되었고 위 제시한 방법이 적합하다는 코멘트 절대 아닙니다!)

junbangg replied:
숨길 수 있는 방법은 어떻게 해야되는지 알것같은데, 최상단으로 올리는 방법은 몰랐습니다!

그런데 혹시 그린이 말씀해주신 "지웠다 생성" 이 메모리에 대한 말씀이실까요??
지금 `listViewController` 와 `gridViewController` 인스턴스는 `MarketViewController` 가 계속 갖고있어서, 매번 생성한다기보다는 `segmentedControl` 을 toggle 할때마다 각 `viewController` 의 `view`를 `superView` 에서 잠깐 떼어주고(`viewController.view.removeFromSuperView()`), 다시 올려놓는 것을 반복하는 방법으로 구현했습니다. 

따라서 각 `viewController`는 물론이고 `view` 의 생성(`loadView()`)도 딱 한번씩만 되는걸로 알고있습니다!

junbangg replied:
저희가 사용한 `removeFromSuperview()` 입니다!
![](https://cdn.discordapp.com/attachments/903510141234716702/932834105282953246/2022-01-18_12.06.42.png)
`view` 를 지우지않고, 그냥 잠깐 떼는 메서드인걸로 이해했습니다

GREENOVER replied:
앗 제 코멘트의 문장이 적절치 않고 혼동을 드렸네요ㅠ
네! 이해하고 계신것처럼 메모리는 아니고 해당 뷰를 떼어주고 다시 올려놓는 방법을 지웠다 생성이라고 모호하게 표현했네요ㅠ 죄송합니다🙇🏻
그 의미로 해석해서 코멘트를 봐주시면 감사하겠습니다🙌

junbangg replied:
아닙니다~~ 혹시 그러면 "해당뷰를 최상단으로 올리는 방법" 은 어떤건가요??
어떻게 하는건지 모르겠습니다 :(

GREENOVER replied:
아하 제가 생각한건 
bringSubview같은 메서드를 생각했었습니다!
그게 현재 코드와 방향에서 더 적절하다 이런건 아니고 그런것도 있는데 이렇게 구현하신게 궁금했었어요🙌
https://developer.apple.com/documentation/uikit/uiview/1622541-bringsubviewtofront

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    private func setupSegmentedControl() {
+        segmentedControl.setTitle("LIST", forSegmentAt: 0)
+        segmentedControl.setTitle("GRID", forSegmentAt: 1)
+        segmentedControl.selectedSegmentTintColor = .systemBlue
```
Green은 어떤가요?ㅋㅋㅋㅋㅋ🙃

na-young-kwon replied:
<img width="367" alt="스크린샷 2022-01-18 오후 12 06 01" src="https://user-images.githubusercontent.com/74536728/149863957-1a4d285e-fd46-4c70-9bf1-b79c7219e6d0.png">
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 나쁘지 않은데요?

GREENOVER replied:
엌ㅋㅋㅋㅋㅋㅋㅋ생각보다 나쁘지 않네욬ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    private func fetchPage(pageNumber: Int, itemsPerPage: Int) {
+        startLoadingIndicator()
+        apiService.fetchPage(pageNumber: pageNumber, itemsPerPage: itemsPerPage) { [weak self] result in
+            self?.stopLoadingIndicator()
```
아마 전체적 PR 질의로 남겨주셨던것 같은데 이 부분보단 132~133라인은 어떨까요? (3번 예시일것 같아요)
실패한다면 얼럿을 구현해주고 131 라인에서 얼럿을 띄워주면 되지 않을까요?
그렇게된다면 데이터 통신 성공/실패때 인디케이터는 스탑되고 그에 따라 정상적으로 뷰가 그려지던가 얼럿이 뜨던가 하면 문제가 없다고 생각해요🙋🏻
또 추가적으로 얼럿이 노출되고 확인같은 버튼을 누르면 다시 재 통신을 하게끔 핸들러를 설정해주면 괜찮다 보이는데
어떻게 생각하시나요?🤔

na-young-kwon replied:
1, 3번 위치 모두 정상적으로 작동하는 것을 확인했어요. 
사실 아직은 큰 차이를 모르겠고, 추후 alert을 구현해보면서 그 차이에 대해서 알게 될 것 같아요.
그때 다시 한 번 그린의 피드백을 참고해볼게요~!

GREENOVER commented on OpenMarket/OpenMarket/Namespaces/CellString.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+enum CellString {
+    static let space = " "
+    static let outOfStock = "품절"
+    static let remainingStock = "잔여수량: "
+}
```
namespace 관리 좋네요!

GREENOVER commented on OpenMarket/OpenMarket/View/GridCell.swift:
```diff
@@ -0,0 +1,103 @@
+        priceLabel.attributedText = nil
+    }
+    
+    func configure(with product: Product) {
+        setImageView(with: product.thumbnailURL)
+        setProductNameLabel(to: product.name)
+        setPriceLabels(with: product.price, and: product.discountedPrice, currency: product.currency)
+        setStockLabel(to: product.stock)
+    }
```
메서드로 설정하는것 아주 좋네요!
객체지향적👍

GREENOVER commented on OpenMarket/OpenMarket/View/GridCell.swift:
```diff
@@ -0,0 +1,103 @@
+    }
+    
+    private func setPriceLabels(with price: Int, and discountedPrice: Int, currency: String) {
+        let dontShowDiscounted = discountedPrice == 0
```
`dontShow`가 `hide`로 대체될 수 있는 네이밍 같습니다🙋🏻

junbangg replied:
저희가 네이밍을 바꾼게 반영이 안된것 같습니다!
할인이 됐다/할인이 안됐다 라는 논리가 먼저 생기고, 그다음에 `label` 들을 적절히 바꿔줘야될것 같아서
`isDiscounted` 라는 네이밍으로 생각했었습니다 원래 :)

```swift
if isDiscounted {
    priceLabel.attributedText = convertToAttributedString(from: priceLabel)
    priceLabel.textColor = .red
    discountedPriceLabel.textColor = .systemGray
} else {
    discountedPriceLabel.isHidden = true
    priceLabel.textColor = .systemGray
}
```

GREENOVER commented on OpenMarket/OpenMarket/View/GridCell.swift:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+    
+    private func convertToAttributedString(from label: UILabel) -> NSMutableAttributedString? {
+        guard let priceLabelString = label.text else {
+            return nil
+        }
+        let attributeString = NSMutableAttributedString(string: priceLabelString)
+        let range = NSRange(location: 0, length: attributeString.length)
+        attributeString.addAttribute(.strikethroughStyle, value: 1, range: range)
+        
+        return attributeString
+    }
```
위 메서드는 Grid와 List 셀에서 공통적으로 중복된다고 보여요🙋🏻
사실 지금처럼 두 셀을 따로 가고 구현해줘도 괜찮지만 어떻게보면 이 메서드 뿐 아니라 설정해주는 많은 메서드들이
전체적으로 다 완전 동일한 구현 중복이 있다고 생각합니다.
하나의 셀 파일에서 Grid, List 조건을 가지고 한번에 구현을 할 수 있을것 같다고 보여요🙋🏻

na-young-kwon replied:
중복된다고 생각하긴 했는데, 하나의 셀 에서 관리한다는 생각은 못했었네요~!

그린의 말씀대로 Grid, List 셀 파일을 없애고 `MarketCell`타입을 만들어서 
하나의 파일에서 로직들을 관리하도록 수정했습니다.

GREENOVER reviewed (approved):
@junbangg @na-young-kwon 
알라딘 조이 두분 너무 고생많으셨습니다!
이번이 정말 쉽지 않은 2주였을텐데 처음 접해보는것도 많았고..!
그런데 아주 훌륭히 그리고 완성도 있게 잘 구현해주셨어요👍
완성도 있는 코드여서 더욱 리뷰하기 편했습니다🙇🏻
***
### 칭찬드리고 싶은 부분
- nameSpace 관리가 아주 좋네요 💯
- 코드가 아주 잘 분리되었으며 객체지향적입니다. 객체간 소통이 너무 좋네요 💯
- 로딩 인디케이터 같은 세심한 구현에 아주 💯
- 회전 시나 여러 디바이스에서도 레이아웃이 잘 잡혔네요 💯
***
### 고민 혹은 신경쓰셨던 부분에 대한 저의 의견
1. 레이블 오토레이아웃에 관한 부분
- 적절히 잘 활용된것 확인했씁니다..!
그런데 하나 궁금한게 있습니다.
![1](https://user-images.githubusercontent.com/72292617/149492866-f1537237-4c7a-4fad-9614-90576a1bd811.png)
위 화면처럼 타이틀 레이블이 길 경우 수량이 나오지 않는데 이건 의도 된건지 궁금합니다🙋🏻
2. 로딩 인디케이터의 애니메이션을 끝내는 시점
- 이 부분도 코드 리뷰 코멘트로 자세히 달긴했습니다.
음.. 지금도 충분히 기능 동작에는 무리가 없지만
추후 얼럿이 구현된다고 볼때는 위치가 저는 3번이 더 나을것 같다고 보였어요🙋🏻
이 부분은 저의 생각이고 다른 의견이 있으시다면 말씀해주세요🙏🏻
3. Cell에 데이터를 전달하는 방법
- 너무 좋아요. 객체지향의 사실과 오해를 스터디 하시나요?ㅋㅋ
아주 그것에 잘 접목되었다고 봅니다.
이에 더 의견달 부분이 없습니다👍
***
### 조언 요청에 대한 저의 의견
1. DeploymentTarget 설정 관련
- 아마 이부분은 4기 채널에서 야곰이 말했던것과 유사합니다. @available을 이용해 14이전과 이후를 분기처리해서 구현할 수 있습니다.
다만 이렇게 된다면 14이전 처리된 코드 분기에서는 `UICollectionLayoutListConfiguration`를 사용하지 못하니 `UICollectionViewFlowLayout`등 다른 방법으로 또 한번 구현이 필요하긴 합니다🥲
이때문에 항상 최신 기술을 쓰는것이 좋은것만은 아니기에 선택을 잘해야할 필요가 있습니다👍
2. String extension 활용
- 활용 너무 좋아요!
둘중에는 저는 아래를 택할거같아요.
이유는 전 비교적 복잡하고 한번 일회성이 아니라면 메서드를 사용할것 같고 뷰에 그냥 단순히 일회성이고 많은 부분의 사용이 없다면
그냥 더 직관적이게 문자열을 합쳐서 쓸것 같습니다..!👍
3. Namespace vs String Extension
- 저는 이경우에는 네임스페이스보다 익스텐션이 더 괜찮아보입니다🙋🏻
공백이기에 특정한 이 상황이 아니고 다른 상황에서도 쓰일 수 있지 않을까해서 확장성 있게 가져가는게 더 낫지 않을까요?🤔
4. MarketAPIService 의존성 주입
- 오 고민 잘해보셨고 결론도 잘 도출하신것 같은데요?
현업에서는 말씀하신것처럼 API를 찔러주고 해주는 서비스 모든것을 의존성을 주입하여 사용합니다.
이에 의존성 관리를 해줄 수 있는 라이브러리인 Swinject를 많이들 이용하는데 한번 시간날때 보시면 좋을것 같습니다~!
https://github.com/Swinject/Swinject
그러나 지금은 하나정도기에 말씀하신데로 의존성을 주입해주는 방법이 더 올바를것 같습니다👍
***
### 고민해볼 부분
1. 오토레이아웃 관련
- 아마 시뮬레이터를 동작시켜보시면 디버깅창에 뷰 관련 에러메시지가 나타나는걸 보셨을 수 있습니다🙋🏻
```
Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
2022-01-14 18:45:48.289060+0900 OpenMarket[50395:1937050] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x600003962f80 UIImageView:0x15285a1f0.width == UIImageView:0x15285a1f0.height   (active)>",
    "<NSLayoutConstraint:0x600003963f70 V:[UIImageView:0x15285a1f0]-(12.5)-|   (active, names: '|':UIView:0x15285c470 )>",
    "<NSLayoutConstraint:0x60000395ff70 V:|-(12.5)-[UIImageView:0x15285a1f0]   (active, names: '|':UIView:0x15285c470 )>",
    "<NSLayoutConstraint:0x600003967d40 UIImageView:0x15285a1f0.width == 0.15*UIView:0x15285c470.width   (active)>",
    "<NSLayoutConstraint:0x60000396c000 H:[UIView:0x15285c470]-(0)-|   (active, names: '|':UIView:0x15285bd20 )>",
    "<NSLayoutConstraint:0x60000396c050 V:[UIView:0x15285c470]-(0)-|   (active, names: '|':UIView:0x15285bd20 )>",
    "<NSLayoutConstraint:0x60000396c0a0 V:|-(0)-[UIView:0x15285c470]   (active, names: '|':UIView:0x15285bd20 )>",
    "<NSLayoutConstraint:0x60000396c0f0 H:|-(0)-[UIView:0x15285c470]   (active, names: '|':UIView:0x15285bd20 )>",
    "<NSLayoutConstraint:0x60000396c2d0 'UIIBSystemGenerated' OpenMarket.ListCell:0x152859c90.leading == UIView:0x15285bd20.leading   (active)>",
    "<NSLayoutConstraint:0x60000396c320 'UIIBSystemGenerated' H:[UIView:0x15285bd20]-(0)-|   (active, names: '|':OpenMarket.ListCell:0x152859c90 )>",
    "<NSLayoutConstraint:0x60000396c370 'UIIBSystemGenerated' OpenMarket.ListCell:0x152859c90.top == UIView:0x15285bd20.top   (active)>",
    "<NSLayoutConstraint:0x60000396c3c0 'UIIBSystemGenerated' V:[UIView:0x15285bd20]-(0)-|   (active, names: '|':OpenMarket.ListCell:0x152859c90 )>",
    "<NSLayoutConstraint:0x60000394ec10 'UIView-Encapsulated-Layout-Height' OpenMarket.ListCell:0x152859c90.height == 44   (active)>",
    "<NSLayoutConstraint:0x60000394ebc0 'UIView-Encapsulated-Layout-Width' OpenMarket.ListCell:0x152859c90.width == 414   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x600003962f80 UIImageView:0x15285a1f0.width == UIImageView:0x15285a1f0.height   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
```
이러한 뷰 레이아웃 관련 메시지를 확인해볼 수 있을거에요!
과연 왜 나는지 어떻게 해결하는지 먼저 한번 고민해보시면 좋겠습니다🤔
정 힘들고 해결이 안되면 도움 요청해주세요ㅎㅎ
2. 통신
- 데이터를 현재 20개까지 밖에 안가져오기에 지금 요구되는 구현사항은 아니지만 추후 리팩토링은 필요해보입니다🙋🏻
***
다시 한번 고생많으셨어요 ㅎㅎ
리뷰를 생각보다 빨리 드린것 같은데 주말은 푹 쉬시고 2주 또 달리셔야 하니 월요일 부터 또 힘내보시죠 ㅎㅎ
수정이 필요하면 수정 후 재요청을 아니면 머지 요청을 해주시면 되겠습니다👍

junbangg committed:
refactor: ListCell의 제품명 label과 재고 label 오토레이아웃 수정 #5

- 재고 label에게 더 높은 우선순위 부여

na-young-kwon committed:
refactor: MarketViewController의존성 주입 및 NameSpace 타입 리팩토링 #5

- sceneDelegate에서 MarketViewController에 의존성을 주입하도록 수정
- 의존성 주입을 위한 setAPIService() 메서드 구현
- IdentifiableView 프로토콜 추가
- Identifier 열거형 삭제
- 의존성 분리를 위해 MarketViewController.apiService의 타입을 APIServicable로 변경

na-young-kwon committed:
refactor: MarketViewController의존성 주입 및 NameSpace 타입 리팩토링 #5

- sceneDelegate에서 MarketViewController에 의존성을 주입하도록 수정
- 의존성 주입을 위한 setAPIService() 메서드 구현
- IdentifiableView 프로토콜 추가
- Identifier 열거형 삭제
- 의존성 분리를 위해 MarketViewController.apiService의 타입을 APIServicable로 변경

na-young-kwon committed:
Merge branch 'step2-new' of https://github.com/junbangg/ios-open-market into step2-new

na-young-kwon committed:
refactor: 피드백반영 #5

- 코딩 컨벤션 준수를 위해 길이가 긴 파라미터 세로로 작성
- 네이밍 통일성을 위해 configure -> setup으로 변경
- gridCell에서 dontShow -> isDiscounted로 변경

##### 제목은 최대 50 글자까지만 입력 ############## -> |

# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

junbangg committed:
refactor: GridCell과 ListCell을 MarketCell 클래스로 통일 #5

- NibIdentifier namespace 타입 구현

na-young-kwon commented:
# 답변

### 타이틀 레이블이 길 경우 수량이 나오지 않는데 이건 의도 된건지 궁금합니다

- 이야기 해본 결과 타이틀 레이블이 길 경우 수량의 정보가 더 중요하다고 생각해서 stockLabel의 compression resistance priority 를 더 높게 수정했습니다. 이제 이름 레이블의 길이가 길어져도 stockLabel의 글씨가 잘리지 않습니다. (사실 이름이 길어질수록 셀의 크기도 길어지면 좋겠다고 생각했는데, 어떻게 구현할지 감이 잘 안오네요ㅜㅜ)
---
### 추가질문) 
### 1. MVC에서 데이터 가공은 도대체 어디서 해야할까요?

생각해본 것

- 모델(부적절해 보임)
- 뷰(데이터를 가공하는 메서드를 만든다/현재)
- Presenter 타입 (근데 이건 MVP 같아서 사용하면 안되는거겠죠..?)

### 2. `root ViewController` 에서 생성자로 의존성을 주입할 수 있는 방법은 없나요 그린...?

`sceneDelegate` 에서 `MarketViewController`를 생성하면서 `MarketApiService`인스턴스를 주입시켜 주려고 했는데,  `initializer` 로 주입시키는 방법을 못찾아서 메서드를 통한 주입을 했습니다. `storyboard` 를 사용하면서 주입하려면 `NScoder` 가 필요해서 안되는것 같더라고요..

---
### **고민해볼 부분**

1. 오토레이아웃 warning message 관련

```swift
Will attempt to recover by breaking constraint**

<NSLayoutConstraint:0x600003b9d810 UIImageView:0x15ca29470.width == UIImageView:0x15ca29470.height   (active)>**
```

이것저것 해보니까, 같은 우선순위(1000)의 `constraint` 들이 여러개 있어서 오토레이아웃이 어떤걸 우선시해야되는지 몰라서 발생한것 같은데 맞을까요??

오토레이아웃  제약들이 중복되어 에러들이 발생한다고 추측했습니다.

그래서 제약 중 하나의 우선순위를 낮추니까 에러 메세지들이 없어졌습니다. (List Cell 에서는 이미지뷰의 topAnchor 제약, Grid Cell에서는 이미지뷰의 높이 제약의 우선순위를 낮추었습니다)

GREENOVER commented:
@junbangg @na-young-kwon 
수정하시느라 고생많으셨어요..!
차후 만약 리뷰 재요청이 필요하시다면 우측 상단
![스크린샷 2022-01-18 오후 1 58 04](https://user-images.githubusercontent.com/72292617/149873512-a6fc5b00-1247-4eab-bcd4-1dc484cf7ae6.png)
저기서 리프레쉬 버튼을 눌러주시면 됩니다👍
***
### 질문 및 고민점에 대한 저의 의견
1. MVC에서 데이터 가공은 도대체 어디서 해야할까요?
- 여기서 궁금한게 먼저 있습니다🙋🏻
데이터를 가공하는데 있어 모델은 왜 부적합한지에 대한 의견이 궁금합니다🙋🏻
2. root ViewController 에서 생성자로 의존성을 주입할 수 있는 방법은 없나요 그린...?
- 그렇네요.
확실히 라이브러리를 따로 쓰지 않고 지금 구현에서는 생성자로는 안될것 같습니다.
음.. 추후 코드적으로 의존성주입을 해주는 그런 컨테이너들을 만들고 여기서 모든걸 관리해주는 방향이 맞다고 생각합니다.
그렇게되면 지금 메서드로 의존성을 준 부분을 한곳에서 같은 개념으로 관리는 할 수 있습니다.
지금 적용은 힘들고 조금 이후의 얘기일것 같아요🤔
지금은 우선 이렇게 진행하는게 제일 베스트네요!
3. 오토레이아웃 warning message 관련
- 정확히 디버깅 에러 메시지를 해보셨네요!
추측하시고 해결하신게 맞습니다..!
디버깅창을 항상 유심히 보시면 답이 나오더라구요👍
***
고생 많으셨어요!
잘 작동하는것 확인했씁니다ㅎㅎ 
이후 스텝에서 수정 및 보완할 부분은 진행해도 될것 같아요!
스텝2는 필수적인 사항 다 구현되었기에 여기서 머지하겠습니다👍

na-young-kwon commented:
### 데이터를 가공하는데 있어 모델은 왜 부적합한지에 대한 의견이 궁금합니다🙋🏻
<img width="608" alt="스크린샷 2022-01-18 오후 3 55 52" src="https://user-images.githubusercontent.com/74536728/149886049-6acb2e4d-8998-484a-b976-df792d0ca73c.png">
저희는 이렇게 Model에서 View에 들어갈 문자열을 가공하는 방법은(위 사진에서는 연산 프로퍼피 부분) 부적절 하다고 생각했어요. 
왜냐하면 `어떻게 보일 것인지`에 대한 문제이기 때문에 model 보단 controller나 View에서 해야한다고 생각하기 때문이에요!

GREENOVER commented:
좋습니다!
여기서 포인트는 어떻게 보일지에 대해서겠네요!
그런 맥락이라면 Controller가 적합할것 같습니다.
뷰는 정말 단순히 요소들을 올려 보여줄 수 있고 변경된 부분에 대해서 감지해 보여주는 즉 뷰에서 로직을 판단하거나 하는게 아닌
정말 단순히 보여줌만 제공해주는게 더 적합하다고 생각해요.
그럼으로 MVC에서는 이 뷰에 어떤걸 올려줄지 뷰컨트롤러에서 가공해서 보여주는게 더 적합할것 같아요.
다만 데이터 가공은 모델에서 해주면 안된다가 되면 안되고 지금처럼 어떻게 보일지에 대해서 가공이 필요한거면 컨트롤러를 사용하면 되겠네요!
근데 이건 스타일 차이지만 저는 저런 가공도 어떻게 보일지의 문제가 어떻게 보면 모델 타입을 가지고 결국 가공을 해주고 그 프로퍼티를 사용하는것이라 모델에서 해줄때도 있습니다!
